src/python/pants/rules/core/binary.py

```diff
@@ -61,14 +84,69 @@ class Binary(Goal):
     return Binary(exit_code=0)
 
 
+# TODO: possibly factor this out. Revisit after porting lint.py, fmt.py, and test.py to use the
+#  Target API.
+def implementations_with_supported_target_types(
+    implementations: Iterable[Type[BinaryImplementation]],
+    *,
+    registered_target_types: RegisteredTargetTypes,
+    union_membership: UnionMembership,
+) -> str:
+    implementations_to_target_types = {
+        implementation: implementation.valid_target_types(
+            registered_target_types.types, union_membership=union_membership
+        )
+        for implementation in implementations
+    }
+    return "\n".join(
+        sorted(
+            f"  * {implementation.__name__}, works with target types: "
+            f"{sorted(target_type.alias for target_type in target_types)}"
+            for implementation, target_types in implementations_to_target_types.items()
+        )
+    )
+
+
 @rule
-async def coordinator_of_binaries(target: HydratedTarget) -> CreatedBinary:
-    binary = await Get[CreatedBinary](BinaryTarget, target.adaptor)
-    return binary
+async def coordinator_of_binaries(
+    wrapped_target: WrappedTarget,
+    union_membership: UnionMembership,
+    registered_target_types: RegisteredTargetTypes,
+) -> CreatedBinary:
+    target = wrapped_target.target
+    binary_implementations: Iterable[Type[BinaryImplementation]] = union_membership.union_rules[
+        BinaryImplementation
+    ]
+    valid_binary_implementations = [
+        binary_implementation
+        for binary_implementation in binary_implementations
+        if binary_implementation.is_valid_target(target)
+    ]
+    if not valid_binary_implementations:
+        all_implementations = implementations_with_supported_target_types(
+            binary_implementations,
+            registered_target_types=registered_target_types,
+            union_membership=union_membership,
+        )
+        raise ValueError(
+            f"None of the registered binary implementations work with {target.address} (target "
+            f"type {repr(target.alias)}). All registered binary implementations:\n\n"
+            f"{all_implementations}."
+        )
+    if len(valid_binary_implementations) > 1:
```

Comment:
Does this pre-empt some less well displayed failure from somewhere else in the pipeline? If not it seems like an artificial restrction. It seems completely reasonable to have a `python_binary` in a BUILD file and 1 rule that uses a `PythonPEXBinary` to extract fields needed to build a PEX from that target and another rule that uses a `PythonIPEXBinary` to extract fields needed to build an iPEX from that target and have both rules run and produce two binaries for the `python_binary` input target.

Reply:
This invariant was previously enforced because we had `UnionRule(BinaryTarget, PythonBinaryTargetAdaptor)` and it's impossible for a `TargetAdaptor` to be more than one specific `TargetAdaptor` type. It either is a `PythonBinaryTargetAdaptor` or it's a `MyCustomPythonBinaryTargetAdaptor`; it can't be both. 

We had a rule that went from `PythonBinaryTargetAdaptor -> CreatedBinary`, so we knew when calling `await Get[CreatedBinary](BinaryTarget, target_adaptor)` that we could, by definition, only have one single rule that would go from `PythonBinaryTargetAdaptor -> CreatedBinary`.

Here, we're opening up the flood gates that now it's possible for multiple binary implementations for any given target. 

--

Sometimes, we like that there may be multiple implementations. For example, with `lint.py`, we depend on that to be able to run, say, 4 linters on the exact same target:

https://github.com/pantsbuild/pants/blob/72c4b93bb75216f62bf222ba4a80a43b924f5d74/src/python/pants/rules/core/lint.py#L100-L105

In contrast, with `./v2 test --debug`, we must only have one target, so we have this validation with `--debug`, but not normally:

https://github.com/pantsbuild/pants/blob/72c4b93bb75216f62bf222ba4a80a43b924f5d74/src/python/pants/rules/core/test.py#L205-L222

--

Here, maybe we are okay with multiple binary implementations? I'm not sure if we'd want to run both `Pex` and `IPex`. Maybe, we'd instead want you to do something like `./v2 binary --implementation='ipex'` to disambiguate (which is the current prototype for `./v2 repl`). I'm not sure. (cc @gshuflin)

For this PR, I think we should punt on it and stick with the prior semantics. But rest assured that it is indeed possible to work with multiple binary implementations for a single target.

Reply:
OK - sg for now.

Reply:
It will be interesting to see whether the "filter targets and implementations, and limit to N" code that goals tend to need to write can be made generic (or perhaps even declarative). But walking the line of premature generalization is always challenging.